autoload/db.vim

```diff
@@ -153,33 +170,103 @@ endfunction
 function! s:filter(url, in, ...) abort
   let prefer_filter = a:0 && a:1
   if db#adapter#supports(a:url, 'input') && !(prefer_filter && db#adapter#supports(a:url, 'filter'))
-    return db#adapter#dispatch(a:url, 'input', a:in)
+    return [db#adapter#dispatch(a:url, 'input', a:in), '']
   endif
   let op = db#adapter#supports(a:url, 'filter') ? 'filter' : 'interactive'
-  return s:shell(db#adapter#dispatch(a:url, op)) . ' ' .
-        \ db#adapter#call(a:url, 'input_flag', [], '< ') . shellescape(a:in)
+  return [db#adapter#dispatch(a:url, op), a:in]
 endfunction
 
-function! s:systemlist(cmd, ...) abort
-  let lines = split(call('system', [s:shell(a:cmd)] + a:000), "\n", 1)
-  if len(lines) && empty(lines[-1])
-    call remove(lines, -1)
+function! s:check_job_running(bang) abort
+  let last_preview_buffer = get(t:, 'db_last_preview_buffer', 0)
+  if !a:bang && last_preview_buffer && !empty(getbufvar(last_preview_buffer, 'db_job_id'))
+    throw 'DB: Query already running for this tab'
   endif
-  return lines
+endfunction
+
+function! s:systemlist_job_cb(data, output, status) abort
+  call extend(a:data.content, a:output)
+  let a:data.status = a:status
 endfunction
 
 function! db#systemlist(cmd, ...) abort
-  let lines = call('s:systemlist', [a:cmd] + a:000)
-  return v:shell_error ? [] : lines
+  let job_result = { 'content': [], 'status': 0 }
+  let job = s:job_run(a:cmd, function('s:systemlist_job_cb', [job_result]), '', get(a:, 1, ''))
+  call s:job_wait(job)
+  if !empty(job_result.status)
+    return []
+  endif
+  return job_result.content
+endfunction
+
+function! s:systemlist_with_err(cmd) abort
```

Comment:
This function _does_ need the blank line removal, just like `s:systemlist()` in the old implementation. I assume that's what you were driving at over in the exit callbacks, but we want those to be the raw form for direct writing as a binary file, where as this needs to be normalized output for feeding to completion, etc.

For the same reason, this is where we want to clean up carriage returns on Windows. The only implementation relied on `system()` to do this automatically. Something like this would work:

```vim
if has('win32')
   call map(lines, { _, v -> substitute(v, "\r$", "", "") })
endif
```

Reply:
`s:systemlist_job_cb()` may make more sense, actually. It needs to cover both this function and the `db#systemlist()` variant.

Reply:
>  I assume that's what you were driving at over in the exit callbacks
Yeah I didn't fully understand the problem. Thought we always have an unnecessary empty line.

Added your suggestion. Thanks for elaborating.